src/compiler/checker.ts

```diff
@@ -4954,11 +4958,15 @@ namespace ts {
         }
 
         function createUnaryTypeMapper(source: Type, target: Type): TypeMapper {
-            return t => t === source ? target : t;
+            const mapper = <TypeMapper>(t => t === source ? target : t);
+            mapper.mapsType = sym => sym.name === source.symbol.name;

```

Comment:
This is the information that mappers previously did not expose -- *which* types they map. This is important because `hasTypeParametersInScope` will now tell its caller whether a type needs to be instantiated by a particular mapper.

This is only partly an optimization, because some of the types for which we now skip instantiation should *never* be instantiated, like the anonymous `typeof ListWrapper` types in the generic static functions in the test.